### ISSUE TITLE
The extension never stops to check my bookmarks

### ISSUE THREAD

**Issue by mentalo-** [cadeyrn/bookmarks-organizer#41]
Hi there,

I use Bookmarks Organizer 1.3.0, on Firefox 54.0 (64 bits), and Ubuntu 16.04 64 bits. I have 2086 bookmarks to manage (ok it's a lot, that's why I need to clear them !).

When I run BO for the first time, I have this result : 
- total : 2086
- checked : 2085
- warning : ~200
- errors : ~100 (don't remember the exacts numbers)

Then, I wait... wait... wait, more than 2 hours, and I have never a report of all the bookmarks. It seems freezed.
So I ran again the ckeck. There, it was blocked after 7 bookmarks checked.
I restart Firefox, and after multiple tests, I have never seen a complete report. A every test, it blocked during the analysis, and the screen stays with the 3 bubbles fading in and out.

I don't know if there is a log file to give to you. At the moment, after more than 20 tests, I cannot have a report of my bookmarks, every time it freezes and never give me the hand again ("check bookmarks" button is always disabled after that).

What can I do to find a solution ? How to tell you how to reproduce this ?

Thanks.

**Comment by cadeyrn** [comment]
Hi,

thanks for the report. It's difficult to reproduce because I don't have this problem and because I don't have access to your Firefox profile. There is a debug option in the add-on's settings  but the debug output will only be shown *after* all bookmarks are checked.

I need to solution to cancel a bookmark check after a few seconds and to log such events. Unfortunately it's not yet implemented. At the moment there is nothing what I can do before this enhancement is implemented. I'll try to find a solution soon but I don't have a timeline for this.

Once this is implemented it should no longer takes hours without result, in the worst case some bookmarks will be skipped. But it's better than the current state. And I hope that a log can provide some infos *why* a bookmark could not be checked.

**Comment by Gitoffthelawn** [comment]
+1 Same issue for me.

Thank you for explaining how the debug option works.  I was wondering why nothing ever appears in the console (except for reflows).

**Comment by rayman89** [comment]
@cadeyrn I have the same problem when checking for duplicates. I would suggest that you test on a profile with 2k+ bookmarks. I think either the issue is the amount of bookmarks or the time it spends waiting for an url to resolve.

**Comment by cadeyrn** [comment]
My profile has more than 5000 bookmarks. No urls have to be resolved to check for duplicates. URLs have only to be resolved for the check for broken bookmarks.

**Comment by Gitoffthelawn** [comment]
Here's what I get:

   - When checking for duplicates, it's fairly quick, and shows results.  However, the overlay image with the three pulsing dots never disappears.  Also, the duplicates are listed as *errors*, when they should probably be listed as *warnings*.

   - When checking for broken bookmarks, the extension never finishes.  Interestingly, it always stalls with 3 bookmarks remaining, unless I reinstall the extension, in which case it stalls with 1 bookmark remaining.  The extension never shows any results.

**Comment by rayman89** [comment]
Can confirm that the duplicates "finishes" but stays with the logo and dots which doesn't allow you to do anything to the duplicates in my case.

In my case the broken bookmarks work properly. I have 2k+ bookmarks as well.

**Comment by Gitoffthelawn** [comment]
I wonder what causes the extension to stall for some installations and not others.  BTW, I ran Places Maintenance before running Bookmarks Organizer, which should clean up any errors, and at the least, perform a SQL VACUUM.

**Comment by PadaV4** [comment]
For me the extension stopped at 
Number of bookmarks: 3979
Checked:3978

In the end i narrowed it down to this bookmark i had.
evetaku.com/blog/2010/11/a-picture-drawn-by-an-eraser/
When i removed it, the extension finished checking the others pretty fast.

When i tried with only this bookmark existing(deleted all others), the extension actually did finish, but only after checking for a really long time. Here is the log i got.

[
  {
    "bookmark": {
      "id": "RhVmhMGlDAbh",
      "parentId": "QXLbmU3J1qjm",
      "title": "A Picture Drawn by an Eraser | EveTaku",
      "url": "http://evetaku.com/blog/2010/11/a-picture-drawn-by-an-eraser/",
      "status": 902
    },
    "method": "HEAD",
    "cause": "fetch-error",
    "response": "NetworkError when attempting to fetch resource."
  },
  {
    "bookmark": {
      "id": "RhVmhMGlDAbh",
      "parentId": "QXLbmU3J1qjm",
      "title": "A Picture Drawn by an Eraser | EveTaku",
      "url": "http://evetaku.com/blog/2010/11/a-picture-drawn-by-an-eraser/",
      "status": 902
    },
    "method": "GET",
    "cause": "fetch-error",
    "response": "NetworkError when attempting to fetch resource."
  }
]

**Comment by StanfordAlan** [comment]
I have this too on my Mac 10.12.6 running Firefox 54.0.1 (64-bit). I have 7570 bookmarks.

My experience is similar to that described by Gitoffthelawn. 

Broken hangs at 89 checked though. Duplicates show 7570 checked, but it continues to check.

I've tried restarting and reinstalling the addon.

**Comment by Gitoffthelawn** [comment]
@cadeyrn What is the status of the fix for this?

**Comment by cadeyrn** [comment]
I am busy with other stuff. I'll release a update for my other add-on New Tab Override soon then I'll return to the Bookmarks Organizer development. Thank you for your patience.

**Comment by Gitoffthelawn** [comment]
@cadeyrn Thanks.  Looking forward to the update.

**Comment by cadeyrn** [comment]
A short status update: I can't report any further progress because I was waiting for the Abort API. The Abort API landed in Firefox Nightly and will be enabled by default in Firefox 57. I hope this new API will help with the problem. This API is the only realistic way to solve this problem because it's not possible to abort a fetch() request without this API. This means that Firefox 57 will be required for a improved version.

**Comment by Gitoffthelawn** [comment]
Firefox 57 will officially be released in 2 days.  Can this be fixed now?

Also, fixing the tag issue is important!

**Comment by cadeyrn** [comment]
I will work on an update after the release of Firefox 57. I don't have an timeline for a release date because I have to test a lot and investigate all the issues. But the plan is to stark working on this really soon.

**Comment by Jan02** [comment]
happens to me too: https://streamable.com/ewiss

**Comment by AlphaHasi** [comment]
Same issue her. I tried to find out, what bookmark causes the problem and activated "debug", but where can I find the debug info? Tried JS console and searched in profile folder but there is no file to find.

The browser secreen is empty beside the animated working-icon and the header with the stuuck info:
Anzahl Lesezeichen: 3291
Überprüft: 3289
Warnungen: 1281
Fehler: 230

Can you mabe just activate printing debug info into the console, so you print called URLs and returned code ore timeout info, so even when it breaks you would see, where it breaks?

**Comment by TristisOris** [comment]
could it happens on sites with redirect, captcha bot check, router authorization popup?
plugin works for me only 1 time in old FF release.

**Comment by XhmikosR** [comment]
This is a duplicate of #23

**Comment by cadeyrn** [comment]
Closed as duplicate of #23.

**Comment by mentalo-** [comment]
So it won't be fixed ?

**Comment by cadeyrn** [comment]
It will be fixed once possible, please see the latest comment in #23. I already have a fix but I can't release the fix until Mozilla has fixed the Firefox crash.

**Comment by cadeyrn** [comment]
@mentalo- Mozilla fixed the crash in Firefox 60 and will probably also fix it for Firefox 59 (waiting for confirmation). I implemented my part of the fix in Bookmarks Organizer 2.0 which will require Firefox 59 or higher. In the case it won't be fixed in Firefox 59 I will provide different code pathes for Firefox 59 and 60+ and this issue will be fixed for users of Firefox 60 and higher.

The exact release date of Bookmarks Organizer 2.0 is still TBD but it should be this week.

**Comment by cadeyrn** [comment]
Bookmarks Organizer 2.0 has been released:
https://addons.mozilla.org/en-US/firefox/addon/bookmarks-organizer/

**Comment by mentalo-** [comment]
It works perfecty with 59.0 ! Congrats !
